packages/flutter/lib/src/material/image_icon.dart

```diff
@@ -0,0 +1,98 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+
+import 'icons.dart';
+import 'icon_button.dart';
+import 'icon_theme.dart';
+import 'icon_theme_data.dart';
+
+/// An icon that comes from an [ImageProvider], e.g. an [AssetImage].
+///
+/// See also:
+///
+///  * [IconButton], for interactive icons.
+///  * [IconTheme], which provides ambient configuration for icons.
+///  * [Icon] and [Icons], for icons from the material design library.
+class ImageIcon extends StatelessWidget {
+  /// Creates an image icon.
+  ///
+  /// The [size] and [color] default to the value given by the current [IconTheme].
+  const ImageIcon(this.image, {
+    Key key,
+    this.size,
+    this.color
+  }) : super(key: key);
+
+  /// The image to display as the icon.
+  ///
+  /// The icon can be null, in which case the widget will render as an empty
+  /// space of the specified [size].
+  final ImageProvider image;
+
+  /// The size of the icon in logical pixels.
+  ///
+  /// Icons occupy a square with width and height equal to size.
+  ///
+  /// Defaults to the current [IconTheme] size, if any. If there is no
+  /// [IconTheme], or it does not specify an explicit size, then it defaults to
+  /// 24.0.
+  final double size;
+
+  /// The color to use when drawing the icon.
+  ///
+  /// Defaults to the current [IconTheme] color, if any. If there is
+  /// no [IconTheme], then it defaults to not recolorizing the image.
+  ///
+  /// The image will additionally be adjusted by the opacity of the current
+  /// [IconTheme], if any.
+  final Color color;
+
+  @override
+  Widget build(BuildContext context) {
+    final IconThemeData iconTheme = const IconThemeData.fallback().merge(IconTheme.of(context));
+
+    final double iconSize = size ?? iconTheme.size;
+
+    if (image == null)
+      return new SizedBox(width: iconSize, height: iconSize);
+
+    final double iconOpacity = iconTheme.opacity;
+    final Color iconColor = color ?? iconTheme.color;
+
+    Widget result = new Image(
+      image: image,
+      width: iconSize,
+      height: iconSize,
+      color: iconColor,
+      fit: ImageFit.scaleDown,
```

Comment:
Icon sizes the picture so it always fits (it sets the font size to the size of the icon). I considered `contain`, but that would also stretch it, which seems bad.

I dunno. I plan to eventually make AssetImage just pick a bitmap of the right size, the same way it picks one of the right scale. We now plumb the data all the way from the IconThemeData to the resolver, it just doesn't use it yet.